### PR TITLE
kevin/ENG-993

### PR DESCRIPTION
**Pull request by kcvan**
## Description
Why did you write this code?
feat(signup): add signup page
## How to Test
- [ ] Visit /work/signup
- [ ] check that errors pop up accordingly
- [ ] sign in with google should work
- [ ] signing in should redirect to /work

Which devices did you test on?
- [x] Chrome on Mac
- [ ] Chrome on PC
- [ ] Firefox on Mac
- [ ] Firefox on PC
- [ ] Safari iPhone
- [ ] Chrome Android

### CODE REVIEW

File: src/pages/signup/Signup.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import { FirebaseConsumer, FirebaseUserProps } from 'HOCs/FirebaseProvider';
+
+import SignupForm from './SignupForm';
+
+const Signup = (props: RouteProps) => (
+  <FirebaseConsumer>
+    {({ loading, user }: FirebaseUserProps) => {
+      if (loading) {
+        return (
+          <Container className="d-flex align-items-center justify-content-center position-relative height-100vh">
```
Comment (kcvan) [inline src/pages/signup/Signup.tsx]: need to use styled components to make reusable containers.  will do this in separate pr

File: src/pages/signup/Signup.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        );
+      }
+
+      if (user) {
+        const state = (props.location && props.location.state) || {};
+        const destination = state.referrer || '/work';
+        return <Redirect to={destination} />;
+      }
+      return (
+        <Container className="d-flex flex-column flex-lg-row align-items-center position-relative height-lg-100vh px-0" fluid>
```
Comment (kcvan) [inline src/pages/signup/Signup.tsx]: here too

File: src/pages/signup/SignupForm.tsx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        {({ errors, touched, setFieldValue, submitForm, isSubmitting }) => (
+          <Form method="POST">
+            <div className="mb-7">
+              <h2 className="h3 text-primary font-weight-normal mb-0">
+                Welcome to <span className="font-weight-semi-bold">Beenest</span>
+              </h2>
+              <p>Fill out the form to get started.</p>
+            </div>
+
+            <Row className="d-flex flex-sm-column flex-md-row">
```
Comment (kcvan) [inline src/pages/signup/SignupForm.tsx]: i was thinking about moving these details to a helper array of objects and map it out to avoid boilerplate bloat

File: src/pages/signup/Signup.tsx (hunk 1 above)
Comment (kcvan) [reply src/pages/signup/Signup.tsx]: addressed as string variable

File: src/pages/signup/Signup.tsx (hunk 2 above)
Comment (kcvan) [reply src/pages/signup/Signup.tsx]: addressed

File: src/pages/signup/SignupForm.tsx (hunk 3 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: actually might be too many variables to cover to cleanly do this at this moment

File: src/components-work/AuthenticationTestimonials/AuthenticationTestimonials.tsx [hunk 4]
```diff
@@ -18,7 +18,7 @@ const items = [
   },
 ];
 
-class LoginTestimonials extends React.Component {
+class AuthenticationTestimonials extends React.Component {
```
Comment (tc) [inline src/components-work/AuthenticationTestimonials/AuthenticationTestimonials.tsx]: we should just make this `GuestTestimonials` or `HostTestimonials`; AuthTestimonials doesn't make sense?

File: src/components/work/AuthenticationTestimonials/AuthenticationTestimonials.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    altText: 'This could be the second testimonial',
+    caption: 'Slide 2',
+  },
+  {
+    altText: 'Sometimes this is the third testimonial',
+    caption: 'Slide 3',
+  },
+];
+
+class AuthenticationTestimonials extends React.Component {
```
Comment (tc) [inline src/components/work/AuthenticationTestimonials/AuthenticationTestimonials.tsx]: can we shorten the prefix to `Auth` instead of `Authentication`? It has the same meaning and a lot shorter to type.

File: src/components-work/AuthenticationTestimonials/AuthenticationTestimonials.tsx (hunk 4 above)
Comment (kcvan) [reply src/components-work/AuthenticationTestimonials/AuthenticationTestimonials.tsx]: wasnt sure if we were actually going to use this testimonial carousel, wiull rename to `UserTestimonials`

File: src/components/work/AuthenticationTestimonials/AuthenticationTestimonials.tsx (hunk 5 above)
Comment (kcvan) [reply src/components/work/AuthenticationTestimonials/AuthenticationTestimonials.tsx]: 👍

File: src/pages/signup/SignupForm.tsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+class SignupForm extends React.Component<SignupProps, State> {
+  readonly state = {
+    providerError: null,
+  };
+
+  render() {
+    return (
+      <Formik
+        initialValues={{
+          signupEmail: '',
```
Comment (tc) [inline src/pages/signup/SignupForm.tsx]: why do we need the signup prefix?

File: src/pages/signup/SignupForm.tsx [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+                  </Label>
+                  <Input
+                    type="text"
+                    name="signupFirstName"
+                    id="signupFirstName"
+                    tag={Field}
+                    onChange={(event: React.FormEvent<HTMLInputElement>) => {
+                      setFieldValue('signupFirstName', event.currentTarget.value);
+
+                      if (this.state.providerError !== null) {
```
Comment (tc) [inline src/pages/signup/SignupForm.tsx]: why is the conditional needed?

File: src/pages/signup/SignupForm.tsx [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+                }}
+                placeholder="********"
+                invalid={!!errors.signupPassword && touched.signupPassword}
+              />
+              <FormFeedback>{errors.signupPassword}</FormFeedback>
+            </FormGroup>
+
+            <FormGroup>
+              <Label for="signupConfirmPassword" className="form-label">
+                Confirm Password
```
Comment (tc) [inline src/pages/signup/SignupForm.tsx]: do we need a confirm password?

File: src/pages/signup/SignupForm.tsx (hunk 6 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: i wanted to has a prefix because ids should be unique throughout the app, so having something like `firstName` would probably show up in more than one form on the app

Comment (tc) [reply src/pages/signup/SignupForm.tsx]: why would this need to be unique throughout the app?

File: src/pages/signup/SignupForm.tsx (hunk 7 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: added the conditional so it wouldnt set state if it was already null to prevent extra set states from leaking

File: src/pages/signup/SignupForm.tsx (hunk 6 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: following convention of classes = reusable and id = unique

https://developer.mozilla.org/en-US/docs/Web/API/Element/id

File: src/pages/signup/SignupForm.tsx (hunk 8 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: i dont think it's bad to have it.  i think users expect/dont mind a confirm password field, i feel it adds a more secure feeling to the site.  we can remove if the design calls for it not being there?

File: src/pages/signup/SignupForm.tsx (hunk 7 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: added the conditional because formik/reactstrap isnt good at handling server errors so we us our own state (vs formiks) to show an error from our server/firebase.  formik will reset the error fields on it's own but since formik has it's own state and validators it isnt hooked up to our server/firebase error state

File: src/pages/signup/Signup.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import SignupForm from './SignupForm';
+
+const Signup = (props: RouteProps) => (
+  <FirebaseConsumer>
+    {({ loading, user }: FirebaseUserProps) => {
+      if (loading) {
+        return (
+          <Container className={VIEWPORT_CENTER_LAYOUT}>
+            <Spinner color="primary" style={{ width: '8rem', height: '8rem' }} type="grow" />
```
Comment (tc) [inline src/pages/signup/Signup.tsx]: can you hide this behind a "Loading" component so it's consistent

File: src/pages/signup/SignupForm.tsx (hunk 6 above)
Comment (tc) [reply src/pages/signup/SignupForm.tsx]: "it must be unique in a document"
I think it's saying the id should be unique to a page not the entire app. if we have to prefix every id, our ids are going to be reallyLong.

File: src/pages/signup/SignupForm.tsx (hunk 7 above)
Comment (tc) [reply src/pages/signup/SignupForm.tsx]: can this be a generic function so each of the inputs can re-use the function?
```
handleChange:
   setFieldValue(htmlId, event.currentTarget.value); 
...
```

File: src/pages/signup/SignupForm.tsx (hunk 6 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: i think it's supposed to be the entire app? getElementById only returns one element that matches the id, where as getElementsByClass returns an array of elements

File: src/pages/signup/SignupForm.tsx (hunk 7 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: 👍

File: src/pages/signup/Signup.tsx (hunk 9 above)
Comment (kcvan) [reply src/pages/signup/Signup.tsx]: 👍

File: src/pages/login/LoginForm.tsx [hunk 10]
```diff
@@ -154,7 +155,7 @@ class LoginForm extends React.Component<LoginProps, State> {
 
   handleSubmit = (values: LoginFormInput, actions: FormikActions<LoginFormInput>) => {
     login(values.loginEmail, values.loginPassword).catch(error => {
-      actions.setErrors({ authenticationError: error.message });
+      actions.setErrors({ authenticationError: getFriendlyErrorMessage(error) });
```
Comment (tc) [inline src/pages/login/LoginForm.tsx]: instead of authenticationError, you can name this `submitError` and it can be re-used as the attribute name for submission errors.

File: src/pages/login/Login.tsx [hunk 11]
```diff
@@ -2,18 +2,20 @@ import * as React from 'react';
...
 import LoginForm from './LoginForm';
-import LoginTestimonals from './LoginTestimonials';
 
 const Login = (props: RouteProps) => (
   <FirebaseConsumer>
     {({ loading, user }: FirebaseUserProps) => {
       if (loading) {
         return (
-          <Container className="d-flex align-items-center justify-content-center position-relative height-100vh">
+          <Container className={VIEWPORT_CENTER_LAYOUT}>
```
Comment (tc) [inline src/pages/login/Login.tsx]: The other way of doing this is `<CenterLayoutContainer>` instead of the css classes in a string so the programmer doesn't have to think whether to use a "Container" or a "Row" or a "div".

Unless the intended behavior is that the classes should be applied independent of the component

File: src/pages/login/Login.tsx [hunk 12]
```diff
@@ -25,9 +27,9 @@ const Login = (props: RouteProps) => (
         return <Redirect to={destination} />;
       }
       return (
-        <Container className="d-flex align-items-center position-relative height-lg-100vh px-0" fluid>
-          <AuthenticationHeader />
-          <LoginTestimonals />
+        <Container className={AUTHENTICATION_CONTAINER} fluid>
```
Comment (tc) [inline src/pages/login/Login.tsx]: Maybe replace all ``AUTHENTICATION`` with `AUTH`?

File: src/pages/login/LoginForm.tsx (hunk 10 above)
Comment (kcvan) [reply src/pages/login/LoginForm.tsx]: 👍

File: src/pages/login/Login.tsx (hunk 12 above)
Comment (kcvan) [reply src/pages/login/Login.tsx]: 👍

File: src/pages/signup/SignupForm.tsx [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+      .then((_) => login(input.email, input.password))
+      .then((_) => !auth.currentUser ? Promise.resolve() : auth.currentUser.sendEmailVerification())
+      .catch((error: Error) => {
+        actions.setErrors({ [SUBMIT_ERROR]: getFriendlyErrorMessage(error) });
+        actions.setSubmitting(false);
+      });
+  };
+
+  handleChange = (htmlId: string, event: React.FormEvent<HTMLInputElement>, setFieldValue: (field: string, value: any) => void) => {
+    setFieldValue(htmlId, event.currentTarget.value);
```
Comment (tc) [inline src/pages/signup/SignupForm.tsx]: can the html id come from the event.currentTarget?

File: src/pages/login/LoginForm.tsx [hunk 14]
```diff
@@ -36,48 +47,47 @@ class LoginForm extends React.Component<LoginProps, State> {
...
+          [EMAIL]: '',
+          [PASSWORD]: '',
+          [SUBMIT_ERROR]: '',
         }}
         validationSchema={LoginSchema}
         onSubmit={this.handleSubmit}
       >
         {({ errors, touched, setFieldValue, submitForm, isSubmitting }) => (
-          <Form className="mt-5" method="POST">
+          <Form method="POST">
```
Comment (tc) [inline src/pages/login/LoginForm.tsx]: is the method="POST" necessary?

Comment (kcvan) [reply src/pages/login/LoginForm.tsx]: yeah if we don't have method="POST" it assumes get request and will append the input into into the url parameters

Comment (tc) [reply src/pages/login/LoginForm.tsx]: 👍

File: src/pages/signup/SignupForm.tsx (hunk 13 above)
Comment (kcvan) [reply src/pages/signup/SignupForm.tsx]: nice was able to get it from the `currentTarget.name`~

File: src/pages/login/LoginForm.tsx (hunk 14 above)
Comment (kcvan) [reply src/pages/login/LoginForm.tsx]: reactstrap's `Form` component is pretty basic so it needs things like this

Comment (tc) [reply src/pages/login/LoginForm.tsx]: Why not use Formik's Form component?

Comment (kcvan) [reply src/pages/login/LoginForm.tsx]: actually reactstrap has a nifty `tag` prop that takes an a component.  was able to alias formik's `Form` as `FormikForm` and added it in as the tag prop

Comment (kcvan) [reply src/pages/login/LoginForm.tsx]: also we're using as much reactstrap as possible so we get the right stylings right off the bat

File: src/pages/signup/Signup.tsx (hunk 9 above)
Comment (kcvan) [reply src/pages/signup/Signup.tsx]: i'll address this in a separate PR
